src/helics/core/mpi/MpiComms.cpp

```diff
@@ -9,38 +9,319 @@ Lawrence Livermore National Laboratory, operated by Lawrence Livermore National
 
 */
 #include "MpiComms.h"
-#include "helics/core/ActionMessage.h"
+#include "../../common/AsioServiceManager.h"
+#include "../ActionMessage.hpp"
 #include <memory>
 
+#include <mpi.h>
+
 namespace helics
 {
-MpiComms::MpiComms (const std::string &brokerTarget, const std::string &localTarget)
-    : CommsInterface (brokerTarget, localTarget)
+
+bool MpiComms::mpiCommsExists = false;
+std::mutex MpiComms::mpiSerialMutex;
+int MpiComms::commRank = -1;
+
+MpiComms::MpiComms ()
 {
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms()" << std::endl;
+    initMPI();
+    mpiCommsExists = true;
+}
+
+MpiComms::MpiComms(const int &brokerRank)
+    : brokerRank(brokerRank)
+{
+    if (mpiCommsExists)
+    {
+        std::cout << "WARNING: MPIComms object already created, unexpected results may occur" << std::endl;
+    }
+    std::cout << "MpiComms(" << std::to_string(brokerRank) << ")" << std::endl;
+    initMPI();
+    if (brokerRank == -1)
+    {
+        setBrokerRank (commRank);
```

Comment:
can commRank still be -1 here?

Reply:
Ranks are >= 0; the only exception would be if something went MPI failed to initialize.